File: typescript-examples/playwright-basics-ts/api/06-click-and-navigate.ts
import { BrowserContext, Page } from "playwright";

/**
 * Click and Navigate
 *
 * Demonstrates clicking elements and handling navigation:
 * - Simple clicks
 * - Waiting after clicks
 * - Navigating through pagination
 */

interface Params {
  pagesToNavigate?: number;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
) {
  await page.goto("https://books.toscrape.com/");
  await page.waitForLoadState("networkidle");

  const pagesToNavigate = params.pagesToNavigate ?? 3;
  const visitedPages: string[] = [page.url()];

  for (let i = 0; i < pagesToNavigate - 1; i++) {
    // Find the "next" button
    const nextButton = page.locator("li.next a");

    // Check if next button exists
    if (!(await nextButton.isVisible())) {
      console.log("No more pages to navigate");
      break;
    }

    // Click and wait for navigation
    await nextButton.click();
    await page.waitForLoadState("networkidle");

    visitedPages.push(page.url());
    console.log(`Navigated to page ${i + 2}: ${page.url()}`);
  }

  // Click on a category link
  const travelCategory = page.locator("a", { hasText: "Travel" });
  if (await travelCategory.isVisible()) {
    await travelCategory.click();
    await page.waitForLoadState("networkidle");
    visitedPages.push(page.url());
  }

  return {
    pagesVisited: visitedPages.length,
    urls: visitedPages,
  };
}

File: python-examples/playwright-python/api/06-click-and-navigate.py
"""
Click and Navigate

Demonstrates clicking elements and navigation:
- Clicking links and buttons
- Pagination through pages
- Using go_back() to return to previous page
"""

from typing import TypedDict

from playwright.async_api import Page


class Params(TypedDict):
    pagesToVisit: int


async def automation(page: Page, params: Params | None = None, **_kwargs):
    pages_to_visit = params.get("pagesToVisit", 3) if params else 3

    await page.goto("https://books.toscrape.com/")
    await page.wait_for_load_state("networkidle")

    visited_urls = [page.url]

    # Navigate through pages using the "next" button
    for _ in range(pages_to_visit - 1):
        next_button = page.locator("li.next a")

        if await next_button.count() > 0:
            await next_button.click()
            await page.wait_for_load_state("networkidle")
            visited_urls.append(page.url)
            print(f"Navigated to page: {page.url}")
        else:
            break

    # Click on a category link
    travel_link = page.locator('a:has-text("Travel")')
    if await travel_link.count() > 0:
        await travel_link.click()
        await page.wait_for_load_state("networkidle")
        visited_urls.append(page.url)

    return {
        "pagesVisited": len(visited_urls),
        "urls": visited_urls,
    }
